### ISSUE TITLE
pareidolia became  a zombie program after a test

### ISSUE THREAD

**Issue by Linhua-Sun** [cmdoret/pareidolia#3]
Hi: 
I am also testing pareidolia based on your suggestion (https://github.com/koszullab/chromosight/issues/47).
This program seems to become a zombie program, no longer further run, but did not stop
I test by this:
```
pareidolia -n 8 \
		-k loops \	
Control_Rep1_010K_SmKr.cool,Control_Rep2_010K_SmKr.cool,Heat_Rep1_010K_SmKr.cool,Heat_Rep2_010K_SmKr.cool \
		control,control,heat,heat \
		test_pareidolia.tsv
Changes will be computed relative to condition: control
```

**Comment by cmdoret** [comment]
Hi @Linhua-Sun 

Thanks for trying it out !
The last official release was quite old, I have cleaned up things a bit and made a new release (v0.3.0).
It seems to work on my machine, could you upgrade and try again ?

**Comment by Linhua-Sun** [comment]
Hi, I have retried with the latest tool. But it seems that another error came. The output is following:

`pareidolia -n 8 -k loops Control_R1_10K.cool,Control_R2_10K.cool,Heat_R1_10K.cool,Heat_R2_10K.cool control,control,heat,heat
test_pareidolia.tsv`

```
Changes will be computed relative to condition: control

Chr1 preprocessed
Chr1 missing bins erased
Chr1 correlation matrices computed
Traceback (most recent call last):
  File "/home/sunlinhua/.local/bin/pareidolia", line 8, in <module>
    sys.exit(pareidolia_cmd())
  File "/data/sunlinhua/software/conda/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/data/sunlinhua/software/conda/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/data/sunlinhua/software/conda/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/data/sunlinhua/software/conda/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/sunlinhua/.local/lib/python3.8/site-packages/pareidolia/cli/__init__.py", line 125, in pareidolia_cmd
    changes = pah.change_detection_pipeline(
  File "/home/sunlinhua/.local/lib/python3.8/site-packages/pareidolia/hic_utils.py", line 420, in change_detection_pipeline
    diff, thresh = detection_matrix(
  File "/home/sunlinhua/.local/lib/python3.8/site-packages/pareidolia/hic_utils.py", line 249, in detection_matrix
    diff.data[snr < 0.8] = 0.0
IndexError: boolean index did not match indexed array along dimension 0; dimension is 587476 but corresponding boolean dimension is 587606
```

**Comment by cmdoret** [comment]
Hi @Linhua-Sun,

I have been working on Pareidolia recently and just released a new version, which addresses the bug you encountered, as well as many other issues. I get pretty decent results on synthetic and experimental data now.

You can upgrade to the latest version using pip, it should work now :)

**Comment by Linhua-Sun** [comment]
Hi @cmdoret 
I'm sorry to disturb you. I have tried the latest version (version 0.4.3). However, it seems to also become a zombie program, no longer further run, but did not stop. And a large amount of memory (~ 200GB) was occupied.

**Comment by cmdoret** [comment]
Wow that is very weird ! Could it be that some subprocesses were killed due to overflowing your RAM ?
How many samples did you give it, and how many contacts / sample ?

**Comment by Linhua-Sun** [comment]
I used two samples with two reps: Control_R1_10K.cool, Control_R2_10K.cool, Heat_R1_10K.cool, Heat_R2_10K.cool. They have 41799160, 115034577, 69302379, and 106241373 counts.

**Comment by cmdoret** [comment]
Strange, I have also run it on the same layout (2 reps, 2 conds) using data from Nora et al., 2017 ([untreated](https://data.4dnucleome.org/experiment-set-replicates/4DNES87HWQAX/) vs [treated](https://data.4dnucleome.org/experiment-set-replicates/4DNES7UKQHOX/)). They are in the range of ~ 100M - 500M contacts / sample and it takes ~45min to run it using <8GB RAM.

It could have something to do with the setup... Are you on windows or using an old python version ? I'm running on linux with python 3.8.

You could also try loading only a small chromosomes using e.g. `-r chr8` and see if it works, but memory should not be an issue normally. I honestly have no clue why it takes so much memory on your machine, but I would try updating python packages just in case.

**Comment by Linhua-Sun** [comment]
I used python 3.6.9 in a linux (centos 7). I also tested `-r chr8`, but it still failed.

**Comment by cmdoret** [comment]
I managed to run pareidolia on the 4 mouse samples from Nora  et al. 2017 using 3 different linux systems, with python 3.6, 3.7 and 3.8. I did not encounter any memory issue (it ran on my laptop). 

I would assume it either has something to do with the system you're using or the data. If you have another computer available, It could be worth trying to run it on a different machine. Otherwise, if you want you could send me your cool files and I'll see if I can reproduce the issue.

**Comment by Linhua-Sun** [comment]
Thanks again. I have also test it in Python 3.6 and 3.9 in a linux.
## About memory issue
I found their score is Floating point number. This may be the cause of memory issue reported previously.

`cooler dump -b Control_Rep1_010K_SmKr.cool`

```
0	0	0.195928	938.197
0	1	0.0793489	493.198
0	2	0.0941511	509.909
0	3	0.0503792	312.817
0	4	0.0176512	113.773
0	5	0.0101473	57.8966
0	6	0.00876741	51.6186
0	7	0.00387957	27.7296
0	8	0.00482291	30.398
0	9	0.00398835	28.1157
```

`cooler dump Control_Rep1_010K.cool`

```
0	0	680
0	1	464
0	2	418
0	3	294
0	4	111
0	5	50
0	6	46
0	7	30
0	8	29
0	9	30
```

## Use cooler file with interger
If I used the cooler file with scores are interger, then new error came.

`pareidolia -n 8 \
> -r Chr1 \
> -k loops \
> Control_Rep1_010K.cool,Control_Rep2_010K.cool,Heat_Rep1_010K.cool,Heat_Rep2_010K.cool \
> control,control,heat,heat \
> test_pareidolia.tsv
```

Their output is a little different with previous output.

```
Loading default parameter for kernel 'loops'...
pearson_thresh: 0.3
min_dist: 20000
max_dist: 2000000
Changes will be computed relative to condition: control
Chr1 preprocessed
Chr1 missing bins erased
Chr1 correlation matrices computed
Traceback (most recent call last):
  File "/home/linhua/.local/bin/pareidolia", line 8, in <module>
    sys.exit(pareidolia_cmd())
  File "/data1/linhua/software/anaconda2/envs/hic/lib/python3.6/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/data1/linhua/software/anaconda2/envs/hic/lib/python3.6/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/data1/linhua/software/anaconda2/envs/hic/lib/python3.6/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/data1/linhua/software/anaconda2/envs/hic/lib/python3.6/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/home/linhua/.local/lib/python3.6/site-packages/pareidolia/cli/__init__.py", line 149, in pareidolia_cmd
    mode=mode,
  File "/home/linhua/.local/lib/python3.6/site-packages/pareidolia/hic_utils.py", line 548, in change_detection_pipeline
    tmp_pos, _ = cud.pick_foci(abs(diff), 0.01, min_size=3)
AttributeError: module 'chromosight.utils.detection' has no attribute 'pick_foci'
```

**Comment by Linhua-Sun** [comment]
I have also test it with ipython.

```
import pareidolia.hic_utils as pah
import chromosight.kernels as ck
pah.change_detection_pipeline(
  ["Control_Rep1_010K.cool", "Control_Rep2_010K.cool", "Heat_Rep1_010K.cool", "Heat_Rep2_010K.cool"],
  ["control", "control", "treatment", "treatment"],
  kernel=ck.loops,
  subsample=True,
  n_cpus=8,
)
```

However, as following:

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-15-1c8439d92811> in <module>
      6   kernel=ck.loops,
      7   subsample=True,
----> 8   n_cpus=8,
      9 )

~/.local/lib/python3.6/site-packages/pareidolia/hic_utils.py in change_detection_pipeline(cool_files, conditions, kernel, bed2d_file, region, max_dist, min_dist, subsample, pearson_thresh, density_thresh, n_cpus, mode)
    468     else:
    469         raise ValueError(
--> 470             "Kernel must either be a valid chromosight pattern name, or a 2D numpy.ndarray of floats"
    471         )
    472     # Associate samples with their conditions

ValueError: Kernel must either be a valid chromosight pattern name, or a 2D numpy.ndarray of floats
```

I will send my data to you by e-mail. 
Thanks again.

**Comment by Linhua-Sun** [comment]
Sorry, I found that your mailbox server doesn't seem to respond to my e-mail. I used `cyril.matthey-doret@pasteur.fr`.

**Comment by cmdoret** [comment]
Hi @Linhua-Sun,

This looks good ! The first error should be solved by updating chromosight (I made a few changes last week), the second should be fixed by replacing `ck.loops` with "loops" (the example was wrong in the readme, I changed it).

Indeed the pixels values of the cool file are expected to be integers. Thanks for the clarification, I should probably raise an error when the program starts, otherwise.
